flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManager.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+import org.apache.flink.util.Preconditions;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * The checkpoint failure manager which centralized manage checkpoint failure processing logic.
+ */
+public class CheckpointFailureManager {
+
+	private final AtomicInteger continuousFailureCounter;
+	private final int tolerableCpFailureNumber;
+
+	public CheckpointFailureManager(int tolerableCpFailureNumber) {
+		Preconditions.checkArgument(tolerableCpFailureNumber > 0,
+			"The tolerable checkpoint failure number must be larger than 0.");
+		this.tolerableCpFailureNumber = tolerableCpFailureNumber;
+		this.continuousFailureCounter = new AtomicInteger(0);
+	}
+
+	/**
+	 * Handle checkpoint exception with a handler callback.
+	 *
+	 * @param exception the checkpoint exception.
+	 * @param callback the handler callback which defines the process logic.
+	 */
+	public void handleCheckpointException(CheckpointException exception, FailureHandlerCallback callback) {
```

Comment:
I am sorry the change of the implementation make you confused. The reason I introduced a callback interface because we must make the "decision" (let job fail or something else). Although we have not distinguished the `Trigger` and `Execution` phase. But there are still two different behavior: `sync`(method call in trigger phase) and `async`(message in execution phase). So I think the `handleCheckpoint` in design doc is not enough. We should track the two phases of failure. What do you think?

Reply:
I will take a look into how it is used into the latest commit.